src/shapes/sdfgrid.cpp

```diff
@@ -562,70 +961,95 @@ class SDFGrid final : public SDF<Float, Spectrum> {
+                        (z + 1) * shape_rcp[0])));
 
                     size_t voxel_index =
                         (x + 0) + (y + 0) * (shape_v[0] - 1) +
                         (z + 0) * (shape_v[0] - 1) * (shape_v[1] - 1);
+
                     voxel_indices[count] = voxel_index;
-                    aabbs[count] = optix::BoundingBox3f(bbox);
+                    aabbs[count]         = BoundingBoxType(bbox);
 
                     count++;
                 }
             }
         }
 
-        jit_free(grid);
+        void *aabbs_ptr           = nullptr;
+        size_t *voxel_indices_ptr = nullptr;
+
+        if constexpr (dr::is_cuda_v<Float>) {
+            // TODO: async memcpy
+            aabbs_ptr =
+                jit_malloc(AllocType::Device, sizeof(BoundingBoxType) * count);
+            jit_memcpy(JitBackend::CUDA, aabbs_ptr, aabbs,
+                       sizeof(BoundingBoxType) * count);
 
-        //TODO: async memcpy
-        void *aabbs_ptr = jit_malloc(AllocType::Device, sizeof(optix::BoundingBox3f) * count);
-        jit_memcpy(JitBackend::CUDA, aabbs_ptr, aabbs, sizeof(optix::BoundingBox3f) * count);
+            voxel_indices_ptr = (size_t *) jit_malloc(AllocType::Device,
+                                                      sizeof(size_t) * count);
+            jit_memcpy(JitBackend::CUDA, voxel_indices_ptr, voxel_indices,
+                       sizeof(size_t) * count);
 
-        void *voxel_indices_ptr = jit_malloc(AllocType::Device, sizeof(size_t) * count);
-        jit_memcpy(JitBackend::CUDA, voxel_indices_ptr, voxel_indices, sizeof(size_t) * count);
+            jit_free(aabbs);
+            jit_free(voxel_indices);
+            jit_free(grid);
 
-        return {aabbs_ptr, voxel_indices_ptr, count};
+        } else {
+            // Our initial allocations were host-visible, so just reuse these
+            aabbs_ptr         = aabbs;
+            voxel_indices_ptr = voxel_indices;
+        }
+
+        return { aabbs_ptr, voxel_indices_ptr, count };
     }
 
     /// Computes the gradient for a specific gradient
-    Vector3f voxel_grad(const Point3f& p, const Point3i& voxel_index) const {
+    Vector3f voxel_grad(const Point3f &p, const Point3i &voxel_index) const {
         auto shape = m_grid_texture.tensor().shape();
-        Vector3f resolution = Vector3f(shape[2] - 1, shape[1] - 1, shape[0] - 1);
-
-        Float f[6];
-        Point3f query;
+        Vector3f resolution =
+            Vector3f(shape[2] - 1, shape[1] - 1, shape[0] - 1);
 
+        InputFloat f[6];
+        InputPoint3f query;
         Point3f voxel_size = 1.f / resolution;
-        Point3f p000 = Point3f(voxel_index) * voxel_size;
+        Point3f p000       = Point3f(voxel_index) * voxel_size;
 
         query = rescale_point(Point3f(p000[0] + voxel_size[0], p[1], p[2]));
-        m_grid_texture.eval(query, &f[0]);
+        m_grid_texture.eval(InputPoint3f(query), &f[0]);
```

Comment:
I don't think all the `InputPoint3f` casts are necessary here, or are they? `rescale_point` should return a `InputPoint3f`.